### PR TITLE
Add script to check and show no-signal preview

### PR DESCRIPTION
**Pull request by aps-sids**
When collectors error, encoders change to "hungry" for feed. In this situation, previews change to a "No signal" image.

### CODE REVIEW

**Comment by mithro** [comment]
@aps-sids I'd love to use this. What do you think is left to finish this off?

**Comment by aps-sids** [comment]
@mithro It needs to be merged with the generate.sh script, removing whole hard-coded stuff.

**Comment by mithro** [comment]
@aps-sids Now you have some time, can you take another look at this?

**Comment by aps-sids** [comment]
pngcrush functionality needs to be added.

**Comment by CarlFK** [comment]
What is pngcrush?

On Wed, Feb 4, 2015 at 1:48 PM, Amanpreet Singh <notifications@github.com>
wrote:

> pngcrush functionality needs to be added.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/timvideos/streaming-system/pull/79#issuecomment-72924927>
> .
>



-- 
Carl K

**Comment by aps-sids** [comment]
Images created by this are >100 KB in size so we need to compress them.
Take a look at existing `generate.sh` script.

**Comment by mithro** [comment]
Just use `subprocess.check_call` to call pngcrush directly.

**Comment by CarlFK** [comment]
I don't see a problem with >100 KB images.  This is the exception to showing a video, so minimal in comparison.

**Comment by mithro** [comment]
@CarlFK this takes over the job of the old preview generation script and instead generates no-signal image when there is no video coming through. Hence it needs to produce small PNGs when the signal is coming through.

File: tools/preview/check_signal.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                ! videoscale \
+                ! video/x-raw,width=300,height=217 \
+                ! pngenc snapshot=true \
+                ! filesink location=/srv/%s/snapshot.png" % (url, channel)
+
+    pipeline = Gst.parse_launch(pipeline)
+    bus = pipeline.get_bus()
+    pipeline.set_state(Gst.State.PLAYING)
+    bus.poll(Gst.MessageType.EOS, Gst.CLOCK_TIME_NONE)
+    pipeline.set_state(Gst.State.NULL)
```
Comment (mithro) [inline tools/preview/check_signal.py]: Maybe add a `#FIXME: This should return true if the preview was generated`

**Comment by aps-sids** [comment]
@mithro This is good to go except that we need to make it threaded. Should the whole thing be made threaded or just the preview generation (gstreamer) part?

**Comment by mithro** [comment]
I'll merge this now. Please another pull request with the threading fixes.

**Pull request merged**